### PR TITLE
[RUNTIME] Simple NDArray container API in c++

### PR DESCRIPTION
**Pull request by tqchen**
Previously TVMArray simply allocates a DLTensor which is a plain C type. This PR change that into a NDArray, which is reference counted and can be used as container in c++.

The TVMArray now is backed by the C++ NDArray::Container, and can pass back and forth in PackedFunc. The python NDArray argument is now passed as NDArray type.

The callee can also directly take out DLTensor* when NDArray is passed, which is compatible with the previous behavior. Moreover, the API allows the callee to take out NDArray from TVMArgs and capture that in the function(by automatically increasing ref count).  PackedFunc can now also return NDArray. 

The NDArray::Container* is DLTensor* compatible, and can be directly viewed as DLTensor*.

## Known Limitations
NDArray is a reference counted type and cannot pass RPC boundary. When NDArray is passed as an argument as a remote function, the argument is downgraded to DLTensor*, so passing NDArray as argument still work if the remote function modifies the data content. Remote function cannot return NDArray.

### CODE REVIEW

File: include/tvm/runtime/ndarray.h [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+  inline void copyfrom(DLTensor* other);
+  inline void copyfrom(const NDArray& other);
+  /*!
+   * \brief Copy data content into another array.
+   * \param other The source array to be copied from.
+   * \note The copy may happen asynchrously if it involves a GPU context.
+   *       TVMSynchronize is necessary.
+   */
+  inline void copyto(DLTensor* other);
+  inline void copyto(const NDArray& other);
```
Comment (jroesch) [inline include/tvm/runtime/ndarray.h]: Do we not use Google style for these? Just want to clarify for example `CopyTo` or `Copyto`?

File: include/tvm/runtime/ndarray.h [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+ private:
+  friend class NDArray;
+  /*!
+   * \brief The shape container,
+   *  can be used used for shape data.
+   */
+  std::vector<int64_t> shape_;
+  /*! \brief The internal array object */
+  std::atomic<int> ref_counter_{0};
```
Comment (jroesch) [inline include/tvm/runtime/ndarray.h]: Do we expect to share these across threads without synchronization? Atomic ref-counts are pretty expensive compared to non-atomic ones.

File: include/tvm/runtime/ndarray.h [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+};
+
+// implementations
```
Comment (jroesch) [inline include/tvm/runtime/ndarray.h]: Improve comment?

File: include/tvm/runtime/packed_func.h [hunk 4]
```diff
@@ -394,8 +409,9 @@ class TVMRetValue : public TVMPODValue_ {
   using TVMPODValue_::operator int;
   using TVMPODValue_::operator bool;
   using TVMPODValue_::operator void*;
-  using TVMPODValue_::operator TVMArray*;
+  using TVMPODValue_::operator DLTensor*;
```
Comment (jroesch) [inline include/tvm/runtime/packed_func.h]: Why do we pass the underlying pointer instead of the container?

File: src/runtime/ndarray.cc [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    using tvm::runtime::NDArray;
+    if (ptr->manager_ctx != nullptr) {
+      static_cast<NDArray::Container*>(ptr->manager_ctx)->DecRef();
+    } else if (ptr->dl_tensor.data != nullptr) {
+      tvm::runtime::DeviceAPI::Get(ptr->dl_tensor.ctx)->FreeDataSpace(
+          ptr->dl_tensor.ctx, ptr->dl_tensor.data);
+    }
+    delete ptr;
+  }
+  // Deleter for NDArray converted from DLPack
```
Comment (jroesch) [inline src/runtime/ndarray.cc]: Just want to understand, this deleter is used from data which is passed from external DLPack arrays not allocated inside of TVM? Could we maybe improve comment here? we could borrow from text of PR, would probably help for future readers of the this code.

File: src/runtime/ndarray.cc [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  static void DLPackDeleter(NDArray::Container* ptr) {
+    DLManagedTensor* tensor = static_cast<DLManagedTensor*>(ptr->manager_ctx);
+    if (tensor->deleter != nullptr) {
+      (*tensor->deleter)(tensor);
+    }
+    delete ptr;
+  }
+  // local create function that
```
Comment (jroesch) [inline src/runtime/ndarray.cc]: Could we improve the comment to be a full sentences, I can do another pass later if you don't have cycles.

i.e "Local create function which allocates tensor metadata but does not allocate space for the data."

File: include/tvm/runtime/ndarray.h [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+   * \param shape The shape of the new array.
+   * \param dtype The data type of the new array.
+   * \param ctx The context of the Array.
+   * \return The created Array
+   */
+  TVM_DLL static NDArray empty(std::vector<int64_t> shape,
+                               DLDataType dtype,
+                               DLContext ctx);
+  /*!
+   * \brief Create a NDArray backed by dlpack tensor.
```
Comment (jroesch) [inline include/tvm/runtime/ndarray.h]: Create a NDArray backed by **a** dlpack tensor.?

File: include/tvm/runtime/ndarray.h (hunk 1 above)
Comment (tqchen) [reply include/tvm/runtime/ndarray.h]: I am debating this. One exception rule here, is that for user facing API, we try to make it as consistent as possible with numpy. Which means function like empty is used in here. But for CopyTo, I think CamelCase is also good

File: include/tvm/runtime/ndarray.h (hunk 2 above)
Comment (tqchen) [reply include/tvm/runtime/ndarray.h]: We might need this for multi-threaded purposes. For most cases, overhead of atomic is fine compared to overhead of other things(computing and python)

File: include/tvm/runtime/packed_func.h (hunk 4 above)
Comment (tqchen) [reply include/tvm/runtime/packed_func.h]: We store the by pointer. But from user's perspective, they pass in containers

File: include/tvm/runtime/ndarray.h (hunk 2 above)
Comment (jroesch) [reply include/tvm/runtime/ndarray.h]: 👍

File: include/tvm/runtime/packed_func.h (hunk 4 above)
Comment (jroesch) [reply include/tvm/runtime/packed_func.h]: I see now, looking at Python code more carefully.

File: include/tvm/runtime/ndarray.h (hunk 1 above)
Comment (tqchen) [reply include/tvm/runtime/ndarray.h]: I switched to CamelCase

**Comment by tqchen** [comment]
@jroesch I addressed all of your comments, can you take another look?

**Comment by jroesch** [comment]
@tqchen seems good to me 👍

**Pull request merged**